File: R/gammaDensities.R
######################################################################
# gammaDensities.R
#
# copyright (c) 1999-2007, Karl W Broman
#
# last modified Apr, 2007
# first written ~Jan, 2001
#
#     This program is free software; you can redistribute it and/or
#     modify it under the terms of the GNU General Public License,
#     version 3, as published by the Free Software Foundation.
# 
#     This program is distributed in the hope that it will be useful,
#     but without any warranty; without even the implied warranty of
#     merchantability or fitness for a particular purpose.  See the GNU
#     General Public License, version 3, for more details.
# 
#     A copy of the GNU General Public License, version 3, is available
#     at http://www.r-project.org/Licenses/GPL-3
#
# Part of the R/xoi package
# Contains: location.given.one, first.given.two, distance.given.two,
#           joint.given.two, xoprob, ioden, firstden, gammacoi,
#           stahlcoi
#
######################################################################

# dist'n of location of XO given exactly one XO 
location.given.one <-
function(nu, L=103, x, n=400, max.conv=25,
         integr.tol=1e-8, max.subd=1000, min.subd=10)
{
  if(nu <= 0) stop("nu should be positive.")

  if(missing(x)) {
    x <- seq(0,L,length=n+1)
    x <- x[-1]-x[2]/2
  }
  if(any(x < 0 || x > L)) {
    x <- x[x >= 0 & x <= L]
    warning("Dropping values outside of [0,L].")
  }
  x <- x/100
  
  result <- .C("location_given_one",
               as.double(nu),
               as.double(x),
               y=as.double(rep(0,length(x))),
               as.integer(length(x)),
               as.double(L/100),
               as.integer(max.conv),
               as.double(integr.tol),
               as.integer(max.subd),
               as.integer(min.subd),
               PACKAGE="xoi")

  data.frame(x=x*100,f=result$y/100)
}


# dist'n of location of first XO given exactly two XOs 
first.given.two <-
function(nu, L=103, x, n=400, max.conv=25,
         integr.tol=1e-8, max.subd=1000, min.subd=10)
{
  if(nu <= 0) stop("nu should be positive.")

  if(missing(x)) {
    x <- seq(0,L,length=n+1)
    x <- x[-1]-x[2]/2
  }
  if(any(x < 0 || x > L)) {
    x <- x[x >= 0 & x <= L]
    warning("Dropping values outside of [0,L].")
  }
  x <- x/100
  
  result <- .C("first_given_two",
               as.double(nu),
               as.double(L/100),
               as.double(x),
               y=as.double(rep(0,length(x))),
               as.integer(length(x)),
               as.integer(max.conv),
               as.double(integr.tol),
               as.integer(max.subd),
               as.integer(min.subd),
               PACKAGE="xoi")

  data.frame(x=x*100, f=result$y/100)
}

  
# dist'n of distance between XOs given two XOs 
distance.given.two <-
function(nu, L=103, x, n=400, max.conv=25,
         integr.tol=1e-8, max.subd=1000, min.subd=10)
{
  if(nu <= 0) stop("nu should be positive.")

  if(missing(x)) {
    x <- seq(0,L,length=n+1)
    x <- x[-1]-x[2]/2
  }
  if(any(x < 0 || x > L)) {
    x <- x[x >= 0 & x <= L]
    warning("Dropping values outside of [0,L].")
  }
  x <- x/100
  
  result <- .C("distance_given_two",
               as.double(nu),
               as.double(L/100),
               as.double(x),
               y=as.double(rep(0,length(x))),
               as.integer(length(x)),
               as.integer(max.conv),
               as.double(integr.tol),
               as.integer(max.subd),
               as.integer(min.subd),
               PACKAGE="xoi")

  data.frame(x=x*100, f=result$y/100)
}

# joint dist'n of locations of XOs given two XOs 
joint.given.two <-
function(nu, L=103, x, y, n=20, max.conv=25,
         integr.tol=1e-8, max.subd=1000, min.subd=10)
{
  if(nu <= 0) stop("nu should be positive.")

  if(missing(x) && missing(y)) { # compute on a grid
    x <- seq(0,L, length=n+1)
    x <- x[-1]-x[2]/2
    y <- x
    x <- rep(x, n)
    y <- rep(y, rep(n,n))
  }
  else if(missing(x)) { 
    x <- seq(0,L,length=n+1)
    x <- x[-1]-x[2]/2
    m <- length(x)
    x <- rep(x, length(y))
    y <- rep(y, rep(m,length(y)))
  }
  else if(missing(y)) {
    y <- seq(0,L,length=n+1)
    y <- y[-1]-y[2]/2
    m <- length(x)
    x <- rep(x, length(y))
    y <- rep(y, rep(m,length(y)))
  }
  else if(length(x) != length(y)) { # both x and y given, but not same length
    m <- length(x)
    x <- rep(x, length(y))
    y <- rep(y, rep(m,length(y)))
  }
    
  if(any(x < 0 | x > L | y < 0 | y > L)) {
    w <- (x >= 0 & x <= L & y>=0 & y <= L)
    x <- x[w]
    y <- y[w]
    warning("Dropping values outside of [0,L].")
  }

  # only include triangle
  if(any(x>y)) {
    w <- (x <= y)
    x <- x[w]
    y <- y[w]
  }

  x <- x/100
  y <- y/100

  result <- .C("joint_given_two",
               as.double(nu),
               as.double(L/100),
               as.double(x),
               as.double(y),
               z=as.double(rep(0,length(x))),
               as.integer(length(x)),
               as.integer(max.conv),
               as.double(integr.tol),
               as.integer(max.subd),
               as.integer(min.subd),
               PACKAGE="xoi")

  data.frame(x=x*100, y=y*100, f=result$z/10000)
}

# calculate probabilities of 0, 1, 2 and >2 XOs 
xoprob <-
function(nu, L=103, max.conv=25,
         integr.tol=1e-8, max.subd=1000, min.subd=10)
{
  if(nu <= 0) stop("nu should be positive.")

  result <- .C("xoprob",
               as.double(nu),
               as.double(L/100),
               prob=as.double(rep(0,4)),
               as.integer(max.conv),
               as.double(integr.tol),
               as.integer(max.subd),
               as.integer(min.subd),
               PACKAGE="xoi")

  result$prob
}

# inter-crossover density 
ioden <- 
function(nu, L=103, x, n=400, max.conv=25)
{
  if(nu <= 0) stop("nu should be positive.")

  if(missing(x)) {
    x <- seq(0,L,length=n+1)
    x <- x[-1]-x[2]/2
  }
  if(any(x < 0)) {
    x <- x[x >= 0]
    warning("Dropping values < 0")
  }
  x <- x/100
  
  result <- .C("ioden",
               as.double(nu),
               as.double(x),
               y=as.double(rep(0,length(x))),
               as.integer(length(x)),
               as.integer(max.conv),
               PACKAGE="xoi")

  y <- result$y

  data.frame(x=x*100, f=y/100)
}

# density for location of first XO
firstden <- 
function(nu, L=103, x, n=400, max.conv=25)
{
  if(nu <= 0) stop("nu should be positive.")

  if(missing(x)) {
    x <- seq(0,L,length=n+1)
    x <- x[-1]-x[2]/2
  }
  if(any(x < 0)) {
    x <- x[x >= 0]
    warning("Dropping values < 0")
  }
  x <- x/100
  
  result <- .C("firstden",
               as.double(nu),
               as.double(x),
               y=as.double(rep(0,length(x))),
               as.integer(length(x)),
               as.integer(max.conv),
               PACKAGE="xoi")

  data.frame(x=x*100, f=result$y/100)
}

# coincidence function for gamma model
gammacoi <-
function(nu, L=103, x, n=400, max.conv=25)
{
  if(nu <= 0) stop("nu should be positive.")

  if(missing(x)) {
    x <- seq(0,L,length=n+1)
    x <- x[-1]-x[2]/2
  }
  if(any(x < 0)) {
    x <- x[x >= 0]
    warning("Dropping values < 0")
  }
  x <- x/100

  result <- .C("GammaCoincidence",
               as.double(nu),
               as.double(x),
               y=as.double(rep(0,length(x))),
               as.integer(length(x)),
               as.integer(max.conv),
               PACKAGE="xoi")

  data.frame(x=x*100, coincidence=result$y)
}  

# coincidence function for gamma model
stahlcoi <-
function(nu, p=0, L=103, x, n=400, max.conv=25)
{
  if(nu <= 0) stop("nu should be positive.")
  if(p < 0 || p > 1) stop("p should be between 0 and 1.")

  if(missing(x)) {
    x <- seq(0,L,length=n+1)
    x <- x[-1]-x[2]/2
  }
  if(any(x < 0)) {
    x <- x[x >= 0]
    warning("Dropping values < 0")
  }
  x <- x/100

  result <- .C("StahlCoincidence",
               as.double(nu),
               as.double(p),
               as.double(x),
               y=as.double(rep(0,length(x))),
               as.integer(length(x)),
               as.integer(max.conv),
               PACKAGE="xoi")

  data.frame(x=x*100, coincidence=result$y)
}  



# end of gammaDensities.R

File: src/GammaDensities.c
/**********************************************************************
 *
 * GammaDensities.c
 *
 * copyright (c) 1998-2007, Karl W Broman
 * 
 * last modified Apr, 2007
 * first written ~Oct, 1998
 *
 *     This program is free software; you can redistribute it and/or
 *     modify it under the terms of the GNU General Public License,
 *     version 3, as published by the Free Software Foundation.
 * 
 *     This program is distributed in the hope that it will be useful,
 *     but without any warranty; without even the implied warranty of
 *     merchantability or fitness for a particular purpose.  See the GNU
 *     General Public License, version 3, for more details.
 * 
 *     A copy of the GNU General Public License, version 3, is available
 *     at http://www.r-project.org/Licenses/GPL-3
 *
 * Part of the R/xoi package
 * Contains: location_given_one, first_given_two, distance_given_two
 *           xoprob, xoprob_sub, xoprob_subsub, ioden, firstden,
 *           GammaCoincidence, StahlCoincidence
 *
 **********************************************************************/

#include <math.h>
#include <stdlib.h>
#include <stdio.h>
#include "GammaS.h"

/* dist'n of location of XO given exactly one XO */
void location_given_one(double *nu, double *x, double *y, int *n, double *L, 
			int *max_conv, double *integr_tol, int *maxsubd,
			int *minsubd) 
{
  int i;
  double denom;
  struct gamma_data info;

  setup_integr_par(*integr_tol, *maxsubd, *minsubd, &(info.integr_info));

  info.nu = *nu;
  info.length = L;
  info.max_conv = *max_conv;

  denom = my_integrate( (void (*)(double *, int, void *))lg1_sub, 
  			(void *)(&info), 0, *L, info.integr_info);

  for(i=0; i<*n; i++) y[i] = x[i];

  lg1_sub(y, *n, &info);
  for(i=0; i<*n; i++) y[i] /= denom;
}

void lg1_sub(double *x, int n, struct gamma_data *info)
{
  int i;

  for(i=0; i<n; i++) 
    x[i] = exp(ll((*info).nu, 1, x[i],                     (*info).max_conv, (*info).integr_info) +
	       ll((*info).nu, 1, (*info).length[0] - x[i], (*info).max_conv, (*info).integr_info));
}


/* dist'n of location of first XO given exactly two XOs */
void first_given_two(double *nu, double *L, double *x,
		     double *y, int *n, int *max_conv, 
		     double *integr_tol, int *maxsubd, int *minsubd)
{
  double temp[2], denom;
  struct gamma_data info;
  int i;

  temp[0] = *L;

  info.max_conv = *max_conv;
  info.nu = *nu;
  info.length = temp;

  setup_integr_par(*integr_tol, *maxsubd, *minsubd, &(info.integr_info));

  denom = my_integrate((void (*)(double *, int, void *))xoprob_sub, 
  		       (void *)(&info), 0.0, *L, info.integr_info);

  for(i=0; i < *n; i++) y[i] = x[i];
  xoprob_sub(y, *n, &info);
  for(i=0; i < *n; i++) y[i] /= denom;
}



/* joint dist'n of XO locations given exactly two XOs */
void joint_given_two(double *nu, double *L, double *x, double *y,
		     double *z, int *n, int *max_conv, 
		     double *integr_tol, int *maxsubd, int *minsubd)
{
  double temp[2], denom;
  struct gamma_data info;
  int i;

  temp[0] = *L;

  info.max_conv = *max_conv;
  info.nu = *nu;
  info.length = temp;

  setup_integr_par(*integr_tol, *maxsubd, *minsubd, &(info.integr_info));

  denom = my_integrate((void (*)(double *, int, void *))xoprob_sub, 
  		       (void *)(&info), 0.0, *L, info.integr_info);

  for(i=0; i < *n; i++) 
    z[i] = exp(ll(*nu, 1, x[i], *max_conv, info.integr_info) +
	       ll(*nu, 0, y[i]-x[i], *max_conv, info.integr_info) +
	       ll(*nu, 1, *L - y[i], *max_conv, info.integr_info))/denom;
}





/* dist'n of distance between XOs given two XOs */
void distance_given_two(double *nu, double *L, double *x,
			double *y, int *n, int *max_conv,
			double *integr_tol, int *maxsubd,
			int *minsubd)
{
  double temp[2], denom;
  struct gamma_data info;
  int i;

  temp[0] = *L;

  info.max_conv = *max_conv;
  info.nu = *nu;
  info.length = temp;

  setup_integr_par(*integr_tol, *maxsubd, *minsubd, &(info.integr_info));

  denom = my_integrate((void (*)(double *, int, void *))xoprob_sub_b, 
  		       (void *)(&info), 0.0, *L, info.integr_info);

  for(i=0; i < *n; i++) y[i] = x[i];
  xoprob_sub_b(y, *n, &info);
  for(i=0; i < *n; i++) y[i] /= denom;
}


void distance_given_two_sub(double *x, int n, struct gamma_data *info)
{
  int i;

  for(i=0; i<n; i++) 
    x[i] = exp(ll((*info).nu, 2, x[i], (*info).max_conv, (*info).integr_info) +
	       ll((*info).nu, 1, (*info).length[0]-x[i], (*info).max_conv, (*info).integr_info));
}


/* calculate probabilities of 0, 1, 2 and >2 XOs */
void xoprob(double *nu, double *L, double *pr, int *max_conv,
	    double *integr_tol, int *maxsubd, int *minsubd)
{
  double temp[2];
  struct gamma_data info;

  temp[0] = *L;

  info.max_conv = *max_conv;
  info.nu = *nu;
  info.length = temp;

  setup_integr_par(*integr_tol, *maxsubd, *minsubd, &(info.integr_info));

  /* pr of 0 XOs */
  pr[0] = exp(ll(*nu, 3, *L, *max_conv, info.integr_info));

  /* pr of 1 XO */
  pr[1] = my_integrate((void (*)(double *, int, void *))lg1_sub,
		       (void *)(&info), 0.0, *L, info.integr_info);

  /* pr of 2 XO */
  pr[2] = my_integrate((void (*)(double *, int, void *))xoprob_sub, 
		       (void *)(&info), 0.0, *L, info.integr_info);

  pr[3] = 1.0 - pr[0] - pr[1] - pr[2];

}

/* calculates un-scaled joint density of locations of two XO given that there are two */
void xoprob_subsub(double *x, int n, struct gamma_data *info)
{
  /* (*info).length[0] is chr length */
  /* (*info).length[1] is location of first XO */
  int i;
  double x0, L;

  L = (*info).length[0];
  x0 = (*info).length[1];

  for(i=0; i<n; i++) 
    x[i] = exp(ll((*info).nu, 1, x0, (*info).max_conv, (*info).integr_info) +
	       ll((*info).nu, 0, x[i], (*info).max_conv, (*info).integr_info) +
	       ll((*info).nu, 1, L-x0-x[i], (*info).max_conv, (*info).integr_info));
}

/* calculates un-scaled joint density of locations of two XO given that there are two */
void xoprob_subsub_b(double *x, int n, struct gamma_data *info)
{
  /* (*info).length[0] is chr length */
  /* (*info).length[1] is distance between the first and second XOs */
  int i;
  double x1, L;

  L = (*info).length[0];
  x1 = (*info).length[1];

  for(i=0; i<n; i++) 
    x[i] = exp(ll((*info).nu, 1, x[i], (*info).max_conv, (*info).integr_info) +
	       ll((*info).nu, 0, x1, (*info).max_conv, (*info).integr_info) +
	       ll((*info).nu, 1, L-x1-x[i], (*info).max_conv, (*info).integr_info));
}


/* calculates distribution of location of first XO given that there are two */
void xoprob_sub(double *x, int n, struct gamma_data *info)
{
  int i;

  for(i=0; i<n; i++) {
    (*info).length[1] = x[i];
    x[i] = my_integrate((void (*)(double *, int, void *))xoprob_subsub,
			(void *)info, 0.0, (*info).length[0]-x[i], (*info).integr_info);
  }
}

/* calculates distribution of location of first XO given that there are two */
void xoprob_sub_b(double *x, int n, struct gamma_data *info)
{
  int i;

  for(i=0; i<n; i++) {
    (*info).length[1] = x[i];
    x[i] = my_integrate((void (*)(double *, int, void *))xoprob_subsub_b,
			(void *)info, 0.0, (*info).length[0]-x[i], (*info).integr_info);
  }
}

/* inter-crossover density */
void ioden(double *nu, double *x, double *y, int *n, int *max_conv)
{
  int i;
  struct integr_data temp;

  for(i=0; i< *n; i++)
    y[i] = exp(ll(*nu, 0, x[i], *max_conv, temp));
}

/* inter-crossover density */
void firstden(double *nu, double *x, double *y, int *n, int *max_conv)
{
  int i;
  struct integr_data temp;

  for(i=0; i< *n; i++)
    y[i] = exp(ll(*nu, 1, x[i], *max_conv, temp));
}


void GammaCoincidence(double *nu, double *x, double *y, int *n,
		      int *max_conv)
{
  int i, j;

  for(i=0; i<*n; i++) {
    for(j=1; j<*max_conv; j++) 
      y[i] += mydgamma(x[i], (double)j * *nu, 2.0 * *nu);
    y[i] /= 2.0;
  }
}


void StahlCoincidence(double *nu, double *p, double *x, double *y, int *n,
		      int *max_conv)
{
  int i, j;

  for(i=0; i<*n; i++) {
    y[i] = 0.0;
    for(j=1; j<*max_conv; j++) 
      y[i] += mydgamma(x[i], (double)j * *nu, 2.0 * (1.0 - *p) * *nu);
    y[i] = (y[i]/2.0) + *p;
  }
}

/* end of GammaDensities.c */
